### PR TITLE
LEST-26 - Implemented more equality matchers

### PR DESCRIPTION
**Pull request by yogwoggf**
Overview:
- Implemented the scaffolds for the new equality matchers.
- Added unit tests for `equality.lua`. 100% code coverage. (toBeNil is an alias, so it isn't in the tests.)

## Questions
- Are these unit tests in good style?
- `toBeInstanceOf` is unclear. Should it act like JavaScript, scanning for multiple parent classes? Or.. should it just scan for the immediate parent class (**currently implemented as such**). In Lua, multiple inheritance is implemented in many ways. I don't think we could offer this functionality because the user's project might have implemented it differently.

This PR will remain as a draft until the questions are answered.

### CODE REVIEW

**Comment by yogwoggf** [comment]
Forgot to mention, all tests are passing on my end.

**Comment by yogwoggf** [comment]
Wait.. wouldn't we need to use Lua assertions for all of the matcher unit tests? Let me know what you want to do with the `matcherPassed` utility. Perhaps it should be split into another file?

File: src/matchers/equality.lua [hunk 1]
```diff
@@ -51,13 +51,41 @@ local function toEqual(ctx, received, expected)
...
+			ctx.inverted and " not " or " "
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeInstanceOf(ctx, received, metatable) end
+local function toBeInstanceOf(ctx, received, metatable)
+	return {
+		pass = deepEqual(getmetatable(received), metatable),
```
Comment (Derpius) [inline src/matchers/equality.lua]: This shouldn’t use deep equality

File: src/matchers/equality.test.lua [hunk 2]
```diff
@@ -0,0 +1,141 @@
+-- This test suite doesn't rely on the equality matchers since they're being tested. Instead this suite uses
+-- basic lua assertions to verify matcher behavior.
```
Comment (Derpius) [inline src/matchers/equality.test.lua]: Useless comment

File: src/matchers/equality.test.lua [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+local function matcherPassed(fn)
+	local passed = pcall(fn)
+	return passed
+end
+
+describe("equality matchers", function()
+	describe("toBe", function()
+		it("should pass on equality", function()
+			assert(matcherPassed(function()
+				expect(2).toBe(2)
```
Comment (Derpius) [inline src/matchers/equality.test.lua]: This is testing too much (in fact it’s a full integration test of the entire expect system + matcher).

And assuming this is *actually* what we wanted (which it isn’t in this case), there’s absolutely no reason to use this assert. The matcher will pass the test if it passes and will fail it if it fails, rendering this code redundant.


What these tests should actually do is import the matchers directly, then assert `pass` and the `message` properties of the returned result. It should do this for both normal and inverted modes.

File: src/matchers/equality.test.lua [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+			end))
+		end)
+
+		it("shouldn't error on inequality", function()
+			assert(not matcherPassed(function()
+				expect(40).toEqual(20)
+			end))
+		end)
+
+		it("should pass on deep equality", function()
```
Comment (Derpius) [inline src/matchers/equality.test.lua:126]: Neither this nor the below test perform equality tests on a deep value, only a shallow table

File: src/matchers/equality.test.lua [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+				expect(testTableOne).toEqual(testTableTwo)
+			end))
+		end)
+	end)
+
+	describe("toBeTruthy", function()
+		it("should pass on truthy values", function()
+			assert(matcherPassed(function()
+				expect(true).toBeTruthy()
+				-- FYI: 1 and 0 are both truthy in Lua unlike other languages.
```
Comment (Derpius) [inline src/matchers/equality.test.lua]: Useless comment

File: src/matchers/equality.test.lua [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+				expect(1).toBeTruthy()
+				expect(0).toBeTruthy()
+				expect("").toBeTruthy()
```
Comment (Derpius) [inline src/matchers/equality.test.lua]: If you’re going to test with multiple types, why not use an exhaustive set?

File: src/matchers/equality.test.lua [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+				expect(false).toBeFalsy()
+			end))
+		end)
+
+		it("shouldn't pass on truthy values", function()
+			assert(not matcherPassed(function()
+				expect(true).toBeFalsy()
+				expect(1).toBeFalsy()
+				expect(0).toBeFalsy()
+				expect("").toBeFalsy()
```
Comment (Derpius) [inline src/matchers/equality.test.lua]: If you’re going to test with multiple types, why not use an exhaustive set?

File: src/matchers/equality.test.lua [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+		end)
+	end)
+
+	describe("toBeInstanceOf", function()
+		local TestClass = {}
+		TestClass.__index = TestClass
+
+		function TestClass.new()
+			return setmetatable({ randomMember = 2 }, TestClass)
+		end
```
Comment (Derpius) [inline src/matchers/equality.test.lua]: This is a lot more code than actually needed for a test like this.

```lua
local Class = {}
local instance = setmetatable({}, Class)
```

File: src/matchers/equality.lua [hunk 9]
```diff
@@ -51,13 +51,41 @@ local function toEqual(ctx, received, expected)
...
+			prettyValue(received),
+			ctx.inverted and " not " or " "
+		),
+	}
+end
 
 ---@type lest.Matcher
-local function toBeInstanceOf(ctx, received, metatable) end
+local function toBeInstanceOf(ctx, received, metatable)
+	return {
```
Comment (Derpius) [inline src/matchers/equality.lua:81]: Could you add an assert for the type of `metatable` here?

Note that this isn’t done in other matchers yet but needs to be added.

**Review by Derpius — CHANGES_REQUESTED** [review]

**Comment by Derpius** [comment]
toBeInstance should just do a direct reference equality check between the metatable of received and the passed metatable.

There is no need to worry about inheritance as Lua doesn’t really have it. If the user of the testing framework needs to repeatedly test their own system for inheritance, they can write a custom matcher (or will be able to in the final product)

**Comment by Derpius** [comment]
Would probably be best to wait for me to do one of the other suite of matchers, so you can use that as reference

**Comment by yogwoggf** [comment]
What the fuck

**Comment by yogwoggf** [comment]
Mmmmmmmmmmmmmm lua 5.1 doesn't coerce the table to a string....... so glad we have these tests in the CI now, should've used `prettyValue` anyways

**Comment by yogwoggf** [comment]
Ok, all tests pass now. The new commits introduce an actual fucken unit test suite for `equality.lua` and adds the requested `assert` for `toBeInstanceOf`.

**Review by Derpius — APPROVED** [review]

**Pull request merged**